Generalization of logp test comparison to facilitate relative and/or absolute comparison

ccaprani opened the pull request:
Following a discussion here https://github.com/pymc-devs/pymc-experimental/pull/84#discussion_r981956590 with @ricardoV94 , there may be cases when it is better to do a relative and/or absolute check on the comparison between the *desired* (SciPy) and *actual* (pymc) values. Currently only an absolute test is done, to `1.5*1e-decimal` [(see here)](https://github.com/numpy/numpy/blob/9c5a2b41cc0ade160a01b7423630f7ee1b8e63dc/numpy/testing/_private/utils.py#L1094) precision.

~~This (small) PR maintains current behaviour (i.e. absolute tolerance), but facilitates future expansion to consider both absolute and/or relative comparison.~~

~~Once/if this PR is accepted, then the `atol` and `rtol` arguments can be propagated out to test functions as a kwarg as `decimal` current is. It could help avoid some of the float32 skipped tests and deal with some other edge cases.~~

- Scope creep means this has been done now!

## Implementation
Change from the not-recommended [assert_almost_equal](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html) to the [assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) from `numpy.testing`. 

Maintain current behaviour by using `atol=decimal` and `rtol=0` in comparing the difference between *actual* and *desired* to `atol + rtol * abs(desired)`.

## Major / Breaking Changes
- None

## Bugfixes / New features
- Facilities a new feature in testing in that future logp could be tested relative and/or absolute.

## Docs / Maintenance
- N/A

ccaprani committed:
Refactor tolerance checks for atol, rtol

* Initial generalization of logp test comparison
* Maintain current behaviour tweak

ricardoV94 commented:
Do you want to try some breaking change, and suggest values that look reasonable as a default? That would be great for maintenance / future devs

ccaprani commented:
Changing from using an atol to an rtol gives the same test results on `test_continuous.py` at least:
```python
rtol=10**(-decimal),
atol=0,#1.5 * 10 ** (-decimal),
```
I haven't got `polygamma` installed, so I get:
```
FAILED test_continuous.py::TestMatchesScipy::test_moyal_logcdf - AssertionError: 
FAILED test_continuous.py::TestMoments::test_polyagamma_moment[1.0-0.0-None-0.25] - RuntimeError: polyagamma package is not installed!
FAILED test_continuous.py::TestMoments::test_polyagamma_moment[1.0-z1-None-expected1] - RuntimeError: polyagamma package is not installed!
FAILED test_continuous.py::TestMoments::test_polyagamma_moment[h2-z2-None-expected2] - RuntimeError: polyagamma package is not installed!
FAILED test_continuous.py::TestMoments::test_polyagamma_moment[h3-z3-size3-expected3] - RuntimeError: polyagamma package is not installed!
========================================================================= 5 failed, 268 passed, 2 skipped, 18 warnings in 37.23s ==========================================================================
```

ricardoV94 commented:
Would probably need to tweak the decimal value, no? And also important to check for float32.

ricardoV94 commented:
I suggest you allow the rtol/atol to be configured already so you can unblock your PR on pymc-experimental.

Then we can follow up with a separate PR where we propose changing the default tolerance for the pre-existing tests.

ricardoV94 commented on pymc/tests/distributions/util.py:
```diff
@@ -317,10 +317,11 @@ def _model_input_dict(model, param_vars, pt):
         pt_d = _model_input_dict(model, param_vars, pt)
         pt_logp = pm.Point(pt_d, model=model)
         pt_ref = pm.Point(pt, filter_model_vars=False, model=model)
-        npt.assert_almost_equal(
```
The same changes should be done to `check_logcdf` and `check_selfconsistency_discrete_logcdf`

ccaprani commented:
@ricardoV94 so an initial refactor is done to now allow both atol and rtol to be specified. 

Given that previous behaviour was an absolute tolerance set at `decimals=6` (just considering float64 for now), then this corresponds to 1e-6 as an absolute precision, so I adopted this as default. However, I moved all `check_logp` and `check_logc` from an absolute to an `rtol` check. I also left a default atol of 1e-10, which is necessary in some cases to get current test behaviour.

I confirmed test behaviour is the same, at least for bound, censored, continuous. Running the suite (and I don't have polyagamma isntalled, so that's 5 fails) gives:
```
50 failed, 2010 passed, 2 skipped, 243 xfailed, 164 warnings in 870.40s (0:14:30)
```

Here are a few observations:
- I moved default tolerances to `select_by_precision` so they can be set in one place only.
- Currently the absolute tolerances for float64 and float32 are set everywhere: suggest this is migrated to avoid duplicating defaults.- 
- I did not adjust tests that did not use `check_logp` or `check_logc` - many functions seem to 'roll their own'
- The float32 checks in many cases were set at *extremely* generous levels (e.g. 1e1!)
- `test_continuous.py::TestMatchesScipy::test_moyal_logcdf` fails, and it's nothing to do with this change.

I think this architecture now allows for a tuning/migration to more flexible and even more stringent test tolerances which can be worked on over time.

ccaprani committed:
Elaborate on tolerances in implementing distributions guide

ccaprani commented:
Hi @ricardoV94 just checking in on this as I see with new merges conflicts are arising.

ricardoV94 commented:
@ccaprani let me see if I can get someone else to weight in :) I'll try to revisit again soon as well!

michaelosthege committed:
Fix find & replace accident

ccaprani committed:
Update atol for platform precision

ccaprani committed:
Merge branch 'pymc-devs:main' into test-metric
